File: Language/C#/Book/tmp/LINQ实战/Chapter06to08.CSharp/Common/SampleClasses/Ch8/Publisher.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace LinqInAction.Chapter06to08.Common.SampleClasses.Ch8
{
  partial class Publisher : System.ComponentModel.IDataErrorInfo
  {
    const string UrlPattern = @"^(ht|f)tp(s?)\:\/\/[0-9a-zA-Z]([-.\w]*[0-9a-zA-Z])*(:(0-9)*)*(\/?)([a-zA-Z0-9\-\.\?\,\'\/\\\+&amp;%\$#_]*)?$";

    partial void OnCreated()
    {
      this.ID = Guid.NewGuid();
      this.Name = string.Empty;
      this.WebSite = string.Empty;
    }
    private string CheckRules(string columnName)
    {
      switch (columnName)
      {
        case "Name":
          if (this.Name != null)
          {
            if (this.Name.Length == 0) { return "Required"; }
            if (this.Name.Length >= 50) { return "Too long"; }
          }
          break;
        case "WebSite":
          if (this.WebSite != null)
          {
            if (this.WebSite.Length >= 200) { return "Too long"; }
            if (!System.Text.RegularExpressions.Regex.IsMatch(this.WebSite, UrlPattern))
            { return "Not a valid URL"; }
          }
          break;
      }
      //All rules are ok, return an empty string
      return string.Empty;
    }

    #region IDataErrorInfo Members

    public string Error
    {
      get
      {
        return CheckRules("Name") +
          CheckRules("WebSite");
      }
    }

    public string this[string columnName]
    {
      get { return CheckRules(columnName); }
    }

    #endregion
  }
}

File: Language/C#/Book/LINQ实战/第06章 LINQ to SQL入门/Chapter06to08.VB/Common/SampleClasses/Ch8/Publisher.vb
Imports System.ComponentModel
Namespace SampleClasses.Ch8
    Public Class Publisher
        Implements IDataErrorInfo

        Const UrlPattern As String = "^(ht|f)tp(s?)\:\/\/[0-9a-zA-Z]([-.\w]*[0-9a-zA-Z])*(:(0-9)*)*(\/?)([a-zA-Z0-9\-\.\?\,\'\/\\\+&amp;%\$#_]*)?$"

        Private Sub OnCreated()
            ID = Guid.NewGuid()
            Name = String.Empty
            WebSite = String.Empty
        End Sub

        Private Function CheckRules(ByVal columnName As String) As String
            Select Case columnName
                Case "Name"
                    If Me.Name IsNot Nothing Then
                        If (Me.Name.Length = 0) Then Return "Required"
                        If (Me.Name.Length >= 50) Then Return "Too long"
                    End If
                Case "WebSite"
                    If Me.WebSite IsNot Nothing Then
                        If (Me.WebSite.Length >= 200) Then Return "Too long"
                        If Not System.Text.RegularExpressions.Regex.IsMatch(Me.WebSite, UrlPattern) Then _
                            Return "Not a valid URL"
                    End If
            End Select

            'All rules are ok, return an empty string
            Return String.Empty
        End Function

#Region "IDataErrorInfo Members"
        Public ReadOnly Property [Error]() As String Implements System.ComponentModel.IDataErrorInfo.Error
            Get
                Return CheckRules("Name") & CheckRules("WebSite")
            End Get
        End Property

        Default Public ReadOnly Property Item(ByVal columnName As String) As String Implements System.ComponentModel.IDataErrorInfo.Item
            Get
                Return CheckRules(columnName)
            End Get
        End Property

#End Region

    End Class
End Namespace